networking/Specs/Networking_module_design_spec.md

```diff
@@ -0,0 +1,197 @@
+
+![](https://i.imgur.com/kbyA5Xf.png)
+
+
+### Publisher - Subscriber Patern:
+Here the **Networking module** will be the publisher and **Content and Processing** modules will be the Subscribers. So Content and Processing Modules will be subscribing the data from the Network Module. When these modules want to subscribe for the data then these module would have to implement ```INotificationHandler``` Interface. And after the creation of ```Communicator``` object , they would have to call ```subscribeForNotifications```method of the ```Communicator``` object by passing the ```INotificationHandler```object as an argument.
+So when the Networking Module receives the data ,it would call the handler provided as an argument.
+
+![](https://i.imgur.com/oeAjq3y.png)
+
+
+
+## Flow
+
+
+ ### Data Outgoing:
+![](https://i.imgur.com/SLk2jEy.png)
+
+1. Initially a message come from either of the modules comes to the networking module via send function.
+2. We will put some headers like data identifier (chat, UI) and the message id 
+3. The data is pushed into the queue.
+4. There will be a thread  listening on the other side of the queue which will dequeue the data from the queue.
+5. The thread will send the data over the network.
+
+
+### Data Incoming:
+
+![](https://i.imgur.com/kFzKIOE.png)
+
+1) A socket will be listening on some port initially
+2) when any data is received , we will identify the message based on the identifier
+3) If the message is UI object message , then we will enqueue into the queue of UI message otherwise we will enqueue into the chat message queue
+4) There will be two threads Listening on each queue which waits for the data in the queues.
+5) If there is any data in the queues then these Listener threads will deque the data from the corresponding queue and call the handler of the module of the  corresponding message.
+
+
+## Interfaces
+
+```Java
+public interface ICommunicator {
+    #Start communication , initalize all the threads 
+    void start();
+    #Stop the Communication ,
+    void stop();
+    #Send the data to the destination IP
+    void send(String destination, String message, String identifier);
+    #Subscribe the module for receiving the data from the network module
+    void subscribeForNotifications(String identifier, INotificationHandler handler);
+    //returns a free port
+    int getPort();
+   
+}
+
+```
+
+
+```Java
+//This interface needs to implemented by the processing and content module
+public interface INotificationHandler {
+    void onMessageReceived(String message);    
+}
+
```

Comment:
@sravand123 Can you please change the interface by following the coding guidelines? It will help us keep consistency in code and spec.